lib/CL/clBuildProgram.c

```diff
@@ -340,6 +341,41 @@ CL_API_SUFFIX__VERSION_1_0
   POCL_MSG_PRINT_INFO("building program with options %s\n",
                        program->compiler_options);
 
+  /* if we're rebuilding the program, release the kernels and reset log/status
+   */
+  if ((program->build_status != CL_BUILD_NONE) || program->num_kernels > 0)
+    {
+      cl_kernel k;
+      for (k = program->kernels; k != NULL; k = k->next)
+        {
+          k->program = NULL;
+          --program->pocl_refcount;
+        }
+      program->kernels = NULL;
+      if (program->num_kernels)
+        {
+          program->operating_on_default_kernels = 1;
```

Comment:
1) This code is only releasing kernels outdated default kernels 
2) building is locked properly i think

There is one problem which you can see around lines 536-550, where the program is unlocked b/c the code of clCreateKernel locks+unlocks the program. This can be fixed by refactoring clCreateKernel or using recursive locks (probably the first one is better).